Fix String.slice with negative values as second argument

teoli2003 opened the pull request:
<!-- Please provide the following information to help us review this PR: -->

> What was wrong/why is this fix needed? (quick summary only)

If the second argument of String.slice() is negative and smaller than the first argument, the page incorrectly stated that the result will be "".

> MDN URL of the main page changed

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice

> Issue number (if there is an associated issue)

Fix #5371

> Anything else that could help us review it

Testing using the live sample on the page.

teoli2003 committed:
Fix String.slice with negative values as second argument

wbamberg commented:
Thank you @teoli2003 ! This bit is hard to understand! What does the last sentence of the previous paragraph mean:

> If it is not undefined, and Number(endIndex) is not positive, an empty string is returned.

What does "it" refer to?

I wonder if we would do better to format this section a bit differently, with each condition in a separate paragraph and more consistent language, something like:

> If `endIndex` is omitted, undefined, or greater than `str.length`, `slice()` extracts to the end of the string.
> 
> If `endIndex` is negative, `slice()` extracts to `str.length` + `endIndex`. For example, if `endIndex` is -3, it is treated as `str.length` - 3.
>
> If `endIndex` is not undefined, and `Number(endIndex)` is not positive, an empty string is returned.

(this seems to be wrong??)
>
> If `endIndex` is specified, is positive or equal to zero, and is not greater than `beginIndex`, then an empty string is returned. For example, `slice(-3, 0)`, or `slice(3, 1)` both return "".

What do you think?

wbamberg reviewed (commented):
Let me know what you think about https://github.com/mdn/content/pull/5372#issuecomment-849846401.

teoli2003 committed:
second attempt after review

teoli2003 commented:
> What do you think?

Yes, it wasn't clear at all. I followed your advice and added a paragraph per cases, this gets rid of the imprecise _it_ (I think the remaining ones are clear).

I added example of each cases so that people who are more example-driven will have a practical example, and I rewrote the last case completely and hope it is understandable now.

I hope it is better now.

wbamberg reviewed (approved):
Yeah, this is much better. Thanks for the updates @teoli2003 !

theonlypwner commented:
While this part is easier to understand than before, it doesn't mention what happens if `Number(endIndex)` is `NaN`.
> If `endIndex` represents a position that is before the one represented by `startIndex`, `slice()` returns `""`.

It's not clear that if `endIndex` is `NaN`, it _represents a position before the one represented by `startIndex`_. It's obvious that an `endIndex` of `-Infinity` _represents a position before the one represented by `startIndex`_ though.

---

Also, the part for `beginIndex` should be updated, since it still reads
> If `beginIndex` is not a number after `Number(beginIndex)`, it is treated as `0`.

I suggested changing it to this (but it should instead be updated to something similar to the `endIndex` section):
> If `Number(beginIndex)` is not positive, `beginIndex` is treated as `0`.
